Some improvements of rootless-Logcat

dajkim76 opened the pull request:
Hi, please review below improvements.

* Add row background (white, and grey background)
* Rename Search option  to Filter option 
  - Filter by tag (case insensitive)
  - Filter by keyword (case insensitive, and RED background on keyword)
  - tag and keyword has history list. 
  - Show auto complete from history 
* Add new Search option
  - Just highlight search word, not filter (case insensitive)
  - YELLOW background on search word
  - auto completion form history
* Add "Delete all" button
* Add row long click menu 
  - Copy text
  - Show pretty JSON (and copy button)

Next time I want to implement Overlay log view.
Thank you.

![screenshot_2017-06-09_164157](https://user-images.githubusercontent.com/2870331/26966128-d5577b70-4d33-11e7-92b9-ffbc31be7987.png)

![screenshot_2017-06-09_171450](https://user-images.githubusercontent.com/2870331/26966958-41c42616-4d37-11e7-9b21-61b1c6db337c.png)

![screenshot_2017-06-09_164210](https://user-images.githubusercontent.com/2870331/26966129-d7bf174c-4d33-11e7-8e8b-e0bc32d2140a.png)

![screenshot_2017-06-09_164220](https://user-images.githubusercontent.com/2870331/26966132-d9254804-4d33-11e7-8c53-f367d648ef05.png)

tananaev commented:
First of all, thanks for pull request.

Here are a few comments that I have so far:

1. You have to provide images in all required resolutions or use vector images.
2. Please simplify image names. For example, `ic_share_variant_grey600_24dp.png` can be renamed to `ic_share`. We don't really have that many assets to be that specific.
3. I think you overuse `NonNull` annotation. In some cases you use `NonNull` and then specify empty string as a default. I think that's wrong. Null value is there to indicate no value.
4. You don't need to put English string into a localized file. Android will automatically do it.
5. Please make string keys more structured and organized. For example, instead of `cancel` you can use something `button_cancel`.

There are quite a few smaller comments like catching generic exceptions, but let's fix major ones first.

dajkim76 commented:
Thank you. I will update soon.

dajkim76 commented:
Please review again.  Thx.

_Sent from my LGE LG-F800L using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

tananaev commented:
Doesn't seem like you fixed all the comments. You haven't fixed at least the first one.

dajkim76 commented:
Please comment line by line on source code  if you need to fix :)

tananaev commented:
It's general comment that apply to multiple places. Can you please carefully read each one of them and fix it. You haven't fixed the second one.

dajkim76 commented:
I renamed ```ic_share_variant_grey600_24dp.png``` to ```ic_share_grey_24dp.png``` style .  Need to change to ```ic_share.png```?

tananaev commented:
That's exactly what I meant.

dajkim76 commented:
Hmm. I'm confused.   I named it similar to an existing file. 
mater branch has already this files  ```ic_vertical_align_bottom_white_24dp.png```, ```ic_sync_white_24dp.png```,  ```ic_search_white_24dp.png ```. etc.  
Why this files is not ```ic_vertical_align_bottom.png``` , ```ic_sync.png```, ```ic_search.png```?

Do you want to simplify all image names include already exist files?

tananaev commented:
Sorry about that. I guess I forgot to rename them. I usually the suffixes. It would be great if you can rename existing ones as well.

dajkim76 commented:
Of course I fixed.

tananaev commented on app/src/main/java/com/tananaev/logcat/Line.java:
```diff
@@ -15,30 +15,56 @@
+    @NonNull
+    private final String content;
+    @Nullable
+    private String lowerContent;
```
Do we really need to have `lowerTag` and `lowerContent`? It basically means that we store double the amount of data required. I think we should remove it and just use original strings.

dajkim76 replied:
lowerTag need to filter by tag case insensitive. 
lowerContent is used by cached value. but is not important. I fixed.

tananaev replied:
You can just do a case-insensitive search instead of storing lower-case strings.

tananaev commented on app/src/main/java/com/tananaev/logcat/Line.java:
```diff
@@ -15,30 +15,56 @@
+            lowerTag = matcher.group(2).toString().trim().toLowerCase();
+        } else {
+            level = 'D';
+            lowerTag = "";
```
There is not really much value in defaults here. I think we should just let Java assign null values.

dajkim76 replied:
done

tananaev commented on app/src/main/java/com/tananaev/logcat/LineAdapter.java:
```diff
@@ -30,7 +43,11 @@
     private List<Line> linesFiltered = new LinkedList<>();
 
-    private String keyword;
+    private String tag = "";
```
I think default should be `null`, not empty string. Applies to all variables here.

dajkim76 replied:
done

tananaev commented on app/src/main/java/com/tananaev/logcat/LineAdapter.java:
```diff
@@ -30,7 +43,11 @@
-    private String keyword;
+    private String tag = "";
+    private String keyword = "";
+    private String lowerKeyword = "";
```
I think we shouldn't store "lower" value here either.

dajkim76 replied:
done

tananaev commented on app/src/main/java/com/tananaev/logcat/LineAdapter.java:
```diff
@@ -30,7 +43,11 @@
+    private String keyword = "";
+    private String lowerKeyword = "";
+    private String searchWord = "";
+    private String lowerSearchWord = "";
```
Same here.

dajkim76 replied:
lowerKeyword , lowerSearchword is used by cached value when viewHolder is updated. 
But is not important. I fixed

tananaev commented on app/src/main/java/com/tananaev/logcat/LineAdapter.java:
```diff
@@ -99,8 +146,38 @@ public LineViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
+
+                int index = 0, found;
+                while ((found = lowerContent.indexOf(lowerKeyword, index)) >= 0) {
+                    spannableText.setSpan(new BackgroundColorSpan(Color.RED), found, found + lowerKeyword.length(), Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
```
Colors should be in the resources so we can style them later if we want.

dajkim76 replied:
done

tananaev commented on app/src/main/java/com/tananaev/logcat/LineAdapter.java:
```diff
@@ -99,8 +146,38 @@ public LineViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
+            if (!TextUtils.isEmpty(lowerSearchWord)) {
+                int index = 0, found;
+                while ((found = lowerContent.indexOf(lowerSearchWord, index)) >= 0) {
+                    spannableText.setSpan(new BackgroundColorSpan(Color.YELLOW), found, found + lowerSearchWord.length(), Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
```
Same here.

dajkim76 replied:
done

tananaev commented on app/src/main/java/com/tananaev/logcat/LineAdapter.java:
```diff
@@ -99,8 +146,38 @@ public LineViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
+            SpannableString spannableText = new SpannableString(text);
+            String lowerContent = item.getLowerContent();
+
+            // filter keyword
```
I don't think there is any value in this comment.

dajkim76 replied:
done 
I removed useless comment

tananaev commented on app/src/main/java/com/tananaev/logcat/LineAdapter.java:
```diff
@@ -99,8 +146,38 @@ public LineViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
+                }
+            }
+
+            // highlight search word
```
I don't think there is any value in this comment. It's pretty obvious what we are doing here.

dajkim76 replied:
done

tananaev commented on app/src/main/java/com/tananaev/logcat/LineAdapter.java:
```diff
@@ -120,4 +197,61 @@ public int getItemCount() {
+            });
+            builder.setPositiveButton(R.string.warning_close, null);
+            builder.show();
+        } catch (Exception ex) {
```
Don't catch generic exceptions.

dajkim76 replied:
done 
Separated to IndexOutOfBoundsException, JSONException

tananaev commented on app/src/main/java/com/tananaev/logcat/view/ViewUtils.java:
```diff
@@ -0,0 +1,57 @@
+                        public void run() {
+                            autoCompleteTextView.showDropDown();
+                        }
+                    }, 100);
```
Create a constant for it. By the way, isn't there any existing solution for autocomplete so we don't need to implement it ourselves?

dajkim76 replied:
I made constant.  
Actually I need history list.  AutoCompleteTextView is simple way to show history list.

tananaev commented on app/src/main/res/values/strings.xml:
```diff
@@ -11,4 +11,14 @@
     <string name="warning_text">For this app to work, you need to enable remote debugging on the phone using following adb command:\n\nadb tcpip 5555</string>
+    <string name="hint_filter_by_tag">Filter by tag</string>
+    <string name="menu_delete_all">Delete all</string>
+    <string name="menu_filter">Filter</string>
```
Please sort string so they are grouped by areas. For example, it makes sense for all `menu_` to be next to each other so it's easier to add/modify/remove anything.

dajkim76 replied:
done

dajkim76 commented:
I used empty string ```""``` to avoid useless null checking or depth or NPE.  It's my style. But coding convention is more important. I removed all initial empty string and added null checking.

tananaev commented:
Everything seems to be fine now. Just one last thing I want to be removed is `lowerTag` from `Line` class. If you are concerned about performance, you can implement a case-insensitive string search. It should be fairly simple instead of converting all strings to lower-case every time.

dajkim76 commented:
if you write below code
```Log.d("Debug Test", "test log");```
```lowerTag``` will be ```debug test```. 
lowerTag is assigned by Regex search result like ```level```. Just one time.

If I enter filter tag with 'DEBUG' or 'TEST',  line is filtered by  lowerTag.contains('debug') or lowerTag.contains('test'). It's need just single contains() called when filter() called, have no performance issue.   

But ```line.getContent().toLowerCase()``` (every time view updated or filter()) maybe have performance or memory issue. I think it's not important yet.  We can consider later that use Apache common langs StringUtils.indexOfIgnoreCase (https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/StringUtils.html#indexOfIgnoreCase-java.lang.CharSequence-java.lang.CharSequence-int-) We need to add external jar lib file.

tananaev commented:
I think Apache utils internally use toUppercase which is basically equivalent. Why can't we just have a method that would do a case-insensitive search? It should be fairly easy to implement and probably it should be fairly easy to find some code on stackoverflow that does it.

dajkim76 commented:
I thought that case-insensitive searching is next time improvement item. 
If you want to improve together this time, I will add case-insensitive searching soon. :)

tananaev commented:
Not sure what made you think it's a next time improvement. I'm fine if it's next time, but what I want is to get rid of storing lower-case strings.

dajkim76 commented:
done

tananaev commented:
Great, I've merged the pull request. I will release it as soon as I have time to test everything myself. Probably in a couple of weeks.

Thanks

dajkim76 commented:
Thank you for review long time.
